File: AWS_GUI_Dev/BAWGUI/ViewModels/MainViewModel.vb
Imports System.ComponentModel
Imports System.Globalization
Imports System.IO
Imports System.Windows.Forms
Imports System.Windows.Forms.VisualStyles
Imports BAWGUI.Results.ViewModels

Public Class MainViewModel
    Inherits ViewModelBase

    Public Sub New()
        _resultsViewModel = New ResultsViewModel()
        _showSettingsWindow = New DelegateCommand(AddressOf ShowSettings, AddressOf CanExecute)
        _openFile = New DelegateCommand(AddressOf OpenFileFunc, AddressOf CanExecute)
        _settingsViewModel = New SettingsViewModel
        _currentView = _resultsViewModel
        _currentViewName = "Settings"
        _toggleResultsSettings = New DelegateCommand(AddressOf _switchView, AddressOf CanExecute)
    End Sub

    Private _settingsWin As SettingsWindow

    Private _showSettingsWindow As ICommand
    Public Property ShowSettingsWindow As ICommand
        Get
            Return _showSettingsWindow
        End Get
        Set(ByVal value As ICommand)
            _showSettingsWindow = value
        End Set
    End Property
    Private Sub ShowSettings()
        Dim isWindowOpen = False
        For Each w In Application.Current.Windows
            If w Is _settingsWin Then
                isWindowOpen = True
                w.Activate()
            End If
        Next
        If Not isWindowOpen Then
            _settingsWin = New SettingsWindow
            _settingsWin.DataContext = SettingsViewModel
            _settingsWin.Show()
        End If
    End Sub
    Private Property _resultsViewModel As ResultsViewModel
    Public Property ResultsViewModel As ResultsViewModel
        Get
            Return _resultsViewModel
        End Get
        Set(value As ResultsViewModel)
            _resultsViewModel = value
        End Set
    End Property


    Private _openFile As ICommand
    Public Property OpenFile As ICommand
        Get
            Return _openFile
        End Get
        Set(ByVal value As ICommand)
            _openFile = value
        End Set
    End Property
    Private Sub OpenFileFunc()
        Dim openDirectoryDialog As New FolderBrowserDialog()
        openDirectoryDialog.Description = "Select the directory that result xml files are located "
        Dim folderName = SettingsViewModel.DetectorConfigure.EventPath
        If Directory.Exists(folderName) Then
            openDirectoryDialog.SelectedPath = folderName
        Else
            openDirectoryDialog.SelectedPath = Environment.CurrentDirectory
        End If
        openDirectoryDialog.ShowNewFolderButton = False
        If (openDirectoryDialog.ShowDialog = DialogResult.OK) Then
            Dim selectedInputFolderInfo As New DirectoryInfo(openDirectoryDialog.SelectedPath)
            Dim filenames = selectedInputFolderInfo.GetFiles
            filenames = filenames.OrderBy(Function(x) x.Name).ToArray
            Dim files = New List(Of String)
            Dim dates = New List(Of String)
            For Each file In filenames
                Dim nameFragment = file.Name.Split(New Char() {".", "_"})
                If file.Extension.ToLower = ".xml" And nameFragment.Length = 3 Then
                    Dim datestr = nameFragment(1)
                    Try
                        Date.ParseExact(datestr, "yyMMdd", CultureInfo.InvariantCulture)
                        dates.Add(datestr)
                        files.Add(file.FullName)
                    Catch ex As Exception
                        If file.Name.ToLower = "eventlist_current.xml" Then
                            files.Add(file.FullName)
                        End If
                    End Try
                End If
            Next
            If Not files.Count = dates.Count Then
                dates.Sort()
                Dim lastDate = dates.LastOrDefault
                dates.Add((Convert.ToInt32(lastDate) + 1).ToString)
            End If
            'dates.Sort()
            'Dim startDate As String
            'Dim endDate As String
            'If Not String.IsNullOrEmpty(SettingsViewModel.DataConfigure.ReaderProperty.DateTimeStart) Then
            '    startDate = Date.Parse(SettingsViewModel.DataConfigure.ReaderProperty.DateTimeStart).ToString("yyMMdd")
            'Else
            '    startDate = dates.FirstOrDefault
            'End If
            'If Not String.IsNullOrEmpty(SettingsViewModel.DataConfigure.ReaderProperty.DateTimeEnd) Then
            '    endDate = Date.Parse(SettingsViewModel.DataConfigure.ReaderProperty.DateTimeEnd).ToString("yyMMdd")
            'Else
            '    endDate = dates.LastOrDefault
            'End If
            Try
                '_resultsViewModel.LoadResults(files, startDate, endDate)
                _resultsViewModel.LoadResults(files, dates)
            Catch ex As Exception
                MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End If
        'Dim openFileDialog As New System.Windows.Forms.OpenFileDialog()
        'openFileDialog.Filter = "xml files (*.xml)|*.xml|All files (*.*)|*.*"

        'If openFileDialog.ShowDialog() = System.Windows.Forms.DialogResult.OK Then
        '    Try
        '        _resultsViewModel.LoadResults(openFileDialog.FileName)
        '    Catch ex As Exception
        '        MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
        '    End Try
        'End If
    End Sub
    Private _currentView As Object
    Public Property CurrentView As Object
        Get
            Return _currentView
        End Get
        Set(ByVal value As Object)
            _currentView = value
            OnPropertyChanged()
        End Set
    End Property
    Private _settingsViewModel As SettingsViewModel
    Public Property SettingsViewModel As SettingsViewModel
        Get
            Return _settingsViewModel
        End Get
        Set(ByVal value As SettingsViewModel)
            _settingsViewModel = value
            OnPropertyChanged()
        End Set
    End Property
    Private _currentViewName As String
    Public Property CurrentViewName As String
        Get
            Return _currentViewName
        End Get
        Set(ByVal value As String)
            _currentViewName = value
            OnPropertyChanged()
        End Set
    End Property
    Private _toggleResultsSettings As ICommand
    Public Property ToggleResultsSettings As ICommand
        Get
            Return _toggleResultsSettings
        End Get
        Set(ByVal value As ICommand)
            _toggleResultsSettings = value
        End Set
    End Property
    Private Sub _switchView(obj As Object)
        If CurrentViewName = "Settings" Then
            CurrentViewName = "Results"
            CurrentView = SettingsViewModel
        Else
            CurrentViewName = "Settings"
            CurrentView = ResultsViewModel
        End If
    End Sub
End Class

File: AWS_GUI_Dev/BAWGUI/ViewModels/MainViewModel.cs
using System;
using BAWGUI.RunMATLAB.ViewModels;
using BAWGUI.Results.ViewModels;
using BAWGUI.Settings.ViewModels;
using System.Windows.Input;
using System.IO;
using BAWGUI.SignalManagement.ViewModels;
using BAWGUI.Utilities;
using System.Windows.Forms;
using System.Collections.Generic;
using VoltageStability.Models;
using VoltageStability.ViewModels;
using BAWGUI.CoordinateMapping.ViewModels;
using ModeMeter.Models;
using ModeMeter.ViewModels;
using BAWGUI.CoordinateMapping.Models;
using DissipationEnergyFlow.ViewModels;
using System.Collections.ObjectModel;

namespace BAWGUI.ViewModels
{
    public class MainViewModel: ViewModelBase
    {
        public MainViewModel()
        {
            _settingsVM = new SettingsViewModel();
            _runMatlabVM = new RunMATLABViewModel();
            _resultsVM = new ResultsViewModel();
            _projectControlVM = new ProjectsControlViewModel();
            CoordsTableVM = new CoordinatesTableViewModel();
            _currentView = _settingsVM;
            MainViewSelected = new RelayCommand(_switchView);
            _projectControlVM.RunSelected += _onRunSelected;
            _signalMgr = SignalManager.Instance;
            //_settingsVM.SaveNewTasl += _settingsVM_SaveNewTasl;
            _settingsVM.SaveNewTask += _projectControlVM.CreateNewTask;
            //_projectControlVM.WriteSettingsConfigFile += _projectControlVM_WriteSettingsConfigFile;
            _runMatlabVM.MatlabRunning += _matlabEngineStatusChanged;
            InspectRawSignal = new RelayCommand(_inpsectRawInputSignals);
            InspectSignalByTimeRange = new RelayCommand(_inpsectAllSignalsByTimeRange);
            SiteMappingVM = new SiteMappingViewModel(CoordsTableVM.SiteCoords);
            //SiteMappingVM.AvailableSites = CoordsTableVM.SiteCoords;
            SiteMappingVM.SignalCoordsMappingVM = new SignalCoordsMappingViewModel(CoordsTableVM.SiteCoords, _signalMgr);
            _settingsVM.DEFAreasChanged += _settingsVM_DEFAreasChanged;
            _projectControlVM.ResultsStoragePathChanged += _projectControlVM_ResultsStoragePathChanged;
        }

        private void _settingsVM_DEFAreasChanged()
        {
            foreach (var dtr in SettingsVM.DetectorConfigure.DetectorList)
            {
                if (dtr is DEFDetectorViewModel)
                {
                    var thisdtr = dtr as DEFDetectorViewModel;
                    SiteMappingVM.DEFAreaSiteMappingVM.Areas = new ObservableCollection<EnergyFlowAreaCoordsMappingViewModel>(thisdtr.Areas.Values);
                    break;
                }
            }
        }

        private void _projectControlVM_ResultsStoragePathChanged(object sender, string e)
        {
            CoordsTableVM.LocationCoordinatesFilePath = e + "\\SiteCoordinatesConfig.xml";
        }

        //private void _settingsVM_SaveNewTasl(ref SettingsViewModel svm)
        //{
        //    throw new NotImplementedException();
        //}

        private void _matlabEngineStatusChanged(object sender, bool e)
        {
            var sd = sender as RunMATLABViewModel;
            if (e)
            {
                ProjectControlVM.IsMatlabEngineRunning = true;
                SettingsVM.IsMatlabEngineRunning = true;
                foreach (var pjt in ProjectControlVM.AWProjects)
                {
                    pjt.IsEnabled = false;
                    //if (pjt.ProjectName != sd.Project.ProjectName)
                    //{
                    //    pjt.IsProjectEnabled = false;
                    //    foreach (var run in pjt.AWRuns)
                    //    {
                    //        run.IsRunEnabled = false;
                    //    }
                    //}
                    //else
                    //{
                    //    foreach (var run in pjt.AWRuns)
                    //    {
                    //        if (run.AWRunName != sd.Run.AWRunName)
                    //        {
                    //            run.IsRunEnabled = false;
                    //        }
                    //    }
                    //}
                }
            }
            else
            {
                ProjectControlVM.IsMatlabEngineRunning = false;
                SettingsVM.IsMatlabEngineRunning = false;
                foreach (var pjt in ProjectControlVM.AWProjects)
                {
                    pjt.IsEnabled = true;
                    foreach (var run in pjt.AWRuns)
                    {
                        run.IsEnabled = true;
                    }
                }
            }
        }

        //private void _projectControlVM_WriteSettingsConfigFile(object sender, string e)
        //{
        //    var configWriter = new ConfigFileWriter(_settingsVM);
        //    try
        //    {
        //        configWriter.WriteXmlConfigFile(e);
        //    }
        //    catch (Exception ex)
        //    {
        //        System.Windows.Forms.MessageBox.Show("Error writing config.xml file!\n" + ex.Message, "Error!", System.Windows.Forms.MessageBoxButtons.OK);
        //    }
        //}

        //private SettingsWindow _settingsWin = new SettingsWindow();
        private ResultsViewModel _resultsVM;
        public ResultsViewModel ResultsVM
        {
            get { return _resultsVM; }
            set
            {
                _resultsVM = value;
                OnPropertyChanged();
            }
        }
        private ViewModelBase _currentView;
        public ViewModelBase CurrentView
        {
            get { return _currentView; }
            set
            {
                _currentView = value;
                OnPropertyChanged();
            }
        }
        private SettingsViewModel _settingsVM;
        public SettingsViewModel SettingsVM
        {
            get { return _settingsVM; }
            set
            {
                _settingsVM = value;
                OnPropertyChanged();
            }
        }
        private RunMATLABViewModel _runMatlabVM;
        public RunMATLABViewModel RunMatlabVM
        {
            get { return _runMatlabVM; }
            set
            {
                _runMatlabVM = value;
                OnPropertyChanged();
            }
        }
        public ICommand MainViewSelected { get; set; }
        private void _switchView(object obj)
        {
            if ((string)obj == "Settings")
            {
                if (CurrentView == _signalMgr)
                {
                    _signalMgr.DeSelectAllPlots(null);
                }
                CurrentView = SettingsVM;
            }
            else if((string)obj == "Results")
            {
                if (CurrentView == _settingsVM)
                {
                    _deselectSettingsVMStep(_settingsVM);
                }
                if (CurrentView == _signalMgr)
                {
                    _signalMgr.DeSelectAllPlots(null);
                }
                CurrentView = ResultsVM;
            }
            else if((string)obj == "Coordinates")
            {
                if (CurrentView == _settingsVM)
                {
                    _deselectSettingsVMStep(_settingsVM);
                }
                if (CurrentView == _signalMgr)
                {
                    _signalMgr.DeSelectAllPlots(null);
                }
                CurrentView = CoordsTableVM;
                //CoordsTableVM.MapVM.GMap.InvalidateVisual();
            }
            else
            {
                if (CurrentView == _settingsVM)
                {
                    _deselectSettingsVMStep(_settingsVM);
                }
                CurrentView = _signalMgr;
            }
        }

        private void _deselectSettingsVMStep(SettingsViewModel settingsVM)
        {
            var tIndex = settingsVM.CurrentTabIndex;
            if (tIndex == 1)
            {
                settingsVM.DeSelectAllDataConfigSteps();
            }
            else if (tIndex == 2)
            {
                settingsVM.DeSelectAllProcessConfigSteps();
            }
            else if (tIndex == 3)
            {
                settingsVM.DeSelectAllPostProcessConfigSteps();
            }
            else if (tIndex == 4 || tIndex == 5)
            {
                settingsVM.DeSelectAllDetectors();
            }
        }

        private ProjectsControlViewModel _projectControlVM;
        public ProjectsControlViewModel ProjectControlVM
        {
            get { return _projectControlVM; }
            set
            {
                _projectControlVM = value;
                OnPropertyChanged();
            }
        }
        private SignalManager _signalMgr;
        public SignalManager SignalMgr
        {
            get { return _signalMgr; }
        }
        private void _onRunSelected(object sender, AWProjectViewModel e)
        {
            if (e != null)
            {
                //SettingsVM = new SettingsViewModel();
                //RunMatlabVM = new RunMATLABViewModel();
                //ResultsVM = new ResultsViewModel();
                //SettingsVM.SaveNewTask += _projectControlVM.CreateNewTask;
                //RunMatlabVM.MatlabRunning += _matlabEngineStatusChanged;
                //if (CurrentView is SettingsViewModel)
                //{
                //    CurrentView = SettingsVM;
                //}
                //else
                //{
                //    CurrentView = ResultsVM;
                //}

                SettingsVM.Project = e.Model;
                SettingsVM.Run = e.SelectedRun;
                if (File.Exists(e.SelectedRun.Model.ConfigFilePath))
                {
                    try
                    {
                        var config = new ReadConfigXml.ConfigFileReader(e.SelectedRun.Model.ConfigFilePath);
                        //clean up the signal manager
                        _signalMgr.cleanUp();
                        //read input data files and generate all the signal objects from the data files and put them in the signal manager.
                        var readingDataSourceSuccess = _signalMgr.AddRawSignals(config.DataConfigure.ReaderProperty.InputFileInfos, config.DataConfigure.ReaderProperty.ExampleTime);
                        //pass signal manager into settings.
                        SettingsVM.SignalMgr = _signalMgr;
                        if (readingDataSourceSuccess)
                        {
                            //read config files
                            SettingsVM.DataConfigure = new DataConfig(config.DataConfigure, _signalMgr);
                            SettingsVM.ProcessConfigure = new ProcessConfig(config.ProcessConfigure, _signalMgr);
                            SettingsVM.PostProcessConfigure = new PostProcessCustomizationConfig(config.PostProcessConfigure, _signalMgr);
                            SettingsVM.DetectorConfigure = new DetectorConfig(config.DetectorConfigure, _signalMgr);
                            var cti = SettingsVM.CurrentTabIndex;
                            if (cti < 2)
                            {
                                SettingsVM.ReverseSignalPassedThroughNameTypeUnit();
                            }
                            SettingsVM.CurrentTabIndex = cti;
                            SettingsVM.CurrentSelectedStep = null;
                            e.SelectedRun.Model.DataFileDirectories = new List<string>();
                            foreach (var info in _signalMgr.FileInfo)
                            {
                                e.SelectedRun.Model.DataFileDirectories.Add(info.FileDirectory);
                            }

                            //read voltage stability settings from config file.
                            var vsDetectors = new VoltageStabilityDetectorGroupReader(e.SelectedRun.Model.ConfigFilePath).GetDetector();
                            //add voltage stability detectors to te detector list in the settings.
                            if (vsDetectors.Count > 0)
                            {
                                SettingsVM.DetectorConfigure.ResultUpdateIntervalVisibility = System.Windows.Visibility.Visible;
                                foreach (var detector in vsDetectors)
                                {
                                    SettingsVM.DetectorConfigure.DetectorList.Add(new VoltageStabilityDetectorViewModel(detector, _signalMgr));
                                }
                            }
                            var modeMeters = new ModeMeterReader(e.SelectedRun.Model.ConfigFilePath).GetDetectors();
                            if (modeMeters.Count > 0)
                            {
                                foreach (var mm in modeMeters)
                                {
                                    SettingsVM.DetectorConfigure.DetectorList.Add(new SmallSignalStabilityToolViewModel(mm, _signalMgr));
                                }
                                ModeMeterXmlWriter.CheckMMDirsStatus(e.SelectedRun.Model, modeMeters);
                            }
                        }
                        else
                        {
                            SettingsVM.DataConfigure.ReaderProperty = new ReaderProperties(config.DataConfigure.ReaderProperty, _signalMgr);
                        }
                        //set up DEF area and detector signals on map
                        var signalSiteMappingConfig = new SignalMappingPlotConfigReader(e.SelectedRun.Model.ConfigFilePath);
                        _signalMgr.DistinctMappingSignal();
                        SiteMappingVM.AvailableSites = CoordsTableVM.SiteCoords;
                        SiteMappingVM.SignalCoordsMappingVM = new SignalCoordsMappingViewModel(CoordsTableVM.SiteCoords, _signalMgr, signalSiteMappingConfig.GetSignalCoordsMappingModel());
                        var DEFDetectorFound = false;
                        foreach (var dtr in SettingsVM.DetectorConfigure.DetectorList)
                        {
                            if (dtr is DEFDetectorViewModel)
                            {
                                var thisdtr = dtr as DEFDetectorViewModel;
                                //var areas = new ObservableCollection<EnergyFlowAreaCoordsMappingViewModel>(thisdtr.Areas.Values);
                                SiteMappingVM.DEFAreaSiteMappingVM.Areas = new ObservableCollection<EnergyFlowAreaCoordsMappingViewModel>(thisdtr.Areas.Values);
                                DEFDetectorFound = true;
                                break;
                            }
                        }
                        if (!DEFDetectorFound)
                        {
                            SiteMappingVM.DEFAreaSiteMappingVM.Areas = new ObservableCollection<EnergyFlowAreaCoordsMappingViewModel>();
                        }
                    }
                    catch (Exception ex)
                    {
                        MessageBox.Show("Error in reading config file.\n" + ex.Message, "Error!", MessageBoxButtons.OK);
                    }
                }
                else
                {
                    MessageBox.Show("No Config.xml file found in the task folder.", "Error!", MessageBoxButtons.OK);
                }
                //need to read signal stability results

                RunMatlabVM.Project = e.Model;
                RunMatlabVM.Run = e.SelectedRun;
                ResultsVM.Project = e.Model;
                ResultsVM.Run = e.SelectedRun;
            }
        }
        public CoordinatesTableViewModel CoordsTableVM { get; set; }
        public SiteMappingViewModel SiteMappingVM { get; set; }
        //public SignalCoordsMappingViewModel SignalCoordsMappingVM { get; set; }
        //private void _checkMMDirsStatus(AWRunViewModel task, List<SmallSignalStabilityTool> modeMeters)
        //{
        //    var eventPath = task.Model.EventPath;
        //    var mm = eventPath + "\\MM";
        //    if (!Directory.Exists(mm))
        //    {
        //        Directory.CreateDirectory(mm);
        //        System.Windows.Forms.MessageBox.Show("Modemeter event subfolder MM was just created since it didn't exist.", "warning!", MessageBoxButtons.OK);
        //    }
        //    foreach (var meter in modeMeters)
        //    {
        //        var meterDir = mm + "\\" + meter.ModeMeterName;
        //        if (!Directory.Exists(meterDir))
        //        {
        //            Directory.CreateDirectory(meterDir);
        //            System.Windows.Forms.MessageBox.Show("Subfolder for mode meter " + meter.ModeMeterName + " was just created since it didn't exist.", "warning!", MessageBoxButtons.OK);
        //        }
        //    }
        //}

        private void _checkMMDirsStatus(AWRunViewModel task)
        {
        }

        //private void _projectControlVM_RunSelected(object sender, AWRunViewModel e)
        //{
        //    throw new NotImplementedException();
        //}
        public ICommand InspectRawSignal { get; set; }
        private void _inpsectRawInputSignals(object obj)
        {
            CurrentView = _signalMgr;
            var info = (InputFileInfoViewModel)obj;
            //if ((info.FileType == Core.Models.DataFileType.piDatabase) && (SettingsVM.DataConfigure.ReaderProperty.Model != null))
            //{

            //}
            //else
            //{
                try
                {
                    _signalMgr.GetRawSignalData(info, SettingsVM.DataConfigure.ReaderProperty.ExampleTime);
                }
                catch (Exception ex)
                {
                    MessageBox.Show("Error retrieving data for viewing from example file. " + ex.Message, "Error!", MessageBoxButtons.OK);
                }
            //}
        }
        public ICommand InspectSignalByTimeRange { get; set; }
        private void _inpsectAllSignalsByTimeRange(object obj)
        {
            CurrentView = _signalMgr;
            var pm = obj as ViewResolvingPlotModel;
            try
            {
                _signalMgr.GetSignalDataByTimeRange(pm, ResultsVM.Run);
            }
            catch (Exception ex)
            {
                MessageBox.Show("Error retrieving data for viewing from matlab retrieve data function. " + ex.Message, "Error!", MessageBoxButtons.OK);
            }
            //foreach (var ax in pm.Axes)
            //{
            //    if (ax.IsHorizontal())
            //    {
            //        var start = DateTime.FromOADate(ax.ActualMinimum).ToString("MM/dd/yyyy HH:mm:ss");
            //        var end = DateTime.FromOADate(ax.ActualMaximum).ToString("MM/dd/yyyy HH:mm:ss");
            //        _signalMgr.GetSignalDataByTimeRange(start, end, ResultsVM.Run);
            //        break;
            //    }
            //}
        }

    }
}
